### PR TITLE
ARROW-16424: [C++] Use Uri to parse substrait ReadRel file path

### PR DESCRIPTION
**Pull request by drin**
A PR to subsume #13390

I tried to add on to #13390, but then I rebased and now it seems really complicated to add to that PR.

This PR primarily uses `Uri` to parse file URIs extracted from `substrait::ReadRel`. This occurs when `FromProto` is called on a `Relation` that is a read type (`ReadRel`). Additionally, to make `FromProto` a bit more readable, I moved the code for this case into a separate function, `FromReadRelation`. This function has the following overall flow:

1. Validate ReadRel
2. Initialize FileFormat from the 1st file URI
3. foreach file URI:
    1. Validate the proto URI
    2. Parse using the Uri class
    3. Validate the Uri (should be a `file:///` path)
    4. Accumulate `FileInfo` instances using `LocalFileSystem`
4. return DeclarationInfo for dataset scan over the file URIs

Validation functions are prefixed with `Check` to match `CheckRelCommon`.

### CODE REVIEW

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 1]
```diff
@@ -52,159 +55,239 @@ Status CheckRelCommon(const RelMessage& rel) {
...
+    if (file_info.type() == fs::FileType::File) {
+      rel_fpaths.push_back(file_info);
+    }
+  }
+
+  return Status::OK();
+}
+
+// Function that implements "FromProto" for a substrait::ReadRel (read relation)
+Result<DeclarationInfo> FromReadRelation(const substrait::ReadRel& rel,
```
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:144]: I am not sure if we should change the function name. I believe `FromProto` is used by all the other relations. If we are going to change that format, better to do it in a separate PR for all relations and keep things uniform.

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 2]
```diff
@@ -52,159 +55,239 @@ Status CheckRelCommon(const RelMessage& rel) {
...
+      case substrait::ReadRel_LocalFiles_FileOrFiles::kArrow:
+        format = std::make_shared<dataset::IpcFileFormat>();
+        break;
+
+      default:
+        // TODO: maybe check for ".feather" or ".arrows"?
         return Status::NotImplemented(
-            "substrait::ReadRel with read_type other than LocalFiles");
+            "unknown file format ",
+            "(see substrait::ReadRel::LocalFiles::FileOrFiles::file_format)");
```
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:191]: rather than unknown, may be `unsupported`.

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 1 above)
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:144]: I think your approach is better, may be we can standardize this across other relations too. But I believe it could be better to do it in another PR. I see these components in general

1. validation
2. excluding what is not supported via checks
3. logic to populate the relational entities
4. may be helper functions like (file discovery, etc)

At the moment we have all these within the switch case which is not very easy to go through. 

Shall we create a separate JIRA and work it there? cc @westonpace

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 3]
```diff
@@ -52,159 +55,239 @@ Status CheckRelCommon(const RelMessage& rel) {
   return Status::OK();
 }
 
-Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet& ext_set,
-                                  const ConversionOptions& conversion_options) {
-  static bool dataset_init = false;
-  if (!dataset_init) {
-    dataset_init = true;
-    dataset::internal::Initialize();
+Status CheckReadRelation(const substrait::ReadRel& rel,
```
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:58]: doubt: at a certain point, these unimplemented once will be implemented and the use of this function wouldn't be very useful. Wouldn't it be better to leave them within the main function where we do the rest of the tasks?

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 4]
```diff
@@ -52,159 +55,239 @@ Status CheckRelCommon(const RelMessage& rel) {
...
-      }
+  return Status::OK();
+}
 
-      if (read.has_projection()) {
-        // NOTE: scan_options->projection is not used by the scanner and thus can't be
-        // used for this
-        return Status::NotImplemented("substrait::ReadRel::projection");
-      }
+Status CheckFileItem(const substrait::ReadRel_LocalFiles_FileOrFiles& file_item) {
```
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:87]:
```suggestion
Status CheckFileItem(const substrait::ReadRel::LocalFiles::FileOrFiles& file_item) {
```

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 5]
```diff
@@ -52,159 +55,239 @@ Status CheckRelCommon(const RelMessage& rel) {
...
+    switch (item.path_type_case()) {
+      case substrait::ReadRel_LocalFiles_FileOrFiles::kUriFile: {
+        files.emplace_back(item_uri.path(), fs::FileType::File);
+        break;
       }
 
-      if (read.local_files().has_advanced_extension()) {
-        return Status::NotImplemented(
-            "substrait::ReadRel::LocalFiles::advanced_extension");
+      case substrait::ReadRel_LocalFiles_FileOrFiles::kUriFolder: {
```
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:232]: may be?

```suggestion
      case substrait::ReadRel::LocalFiles::FileOrFiles::kUriFolder: {
```

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 6]
```diff
@@ -52,159 +55,239 @@ Status CheckRelCommon(const RelMessage& rel) {
...
-        } else if (item.path_type_case() ==
-                   substrait::ReadRel_LocalFiles_FileOrFiles::kUriFile) {
-          path = item.uri_file();
-        } else if (item.path_type_case() ==
-                   substrait::ReadRel_LocalFiles_FileOrFiles::kUriFolder) {
-          path = item.uri_folder();
-        } else {
-          path = item.uri_path_glob();
-        }
+      case substrait::ReadRel_LocalFiles_FileOrFiles::kUriPath: {
```
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:237]: may be?

```suggestion
      case substrait::ReadRel::LocalFiles::FileOrFiles::kUriPath: {
```

**Comment by vibhatha** [comment]
@drin I added a few comments and suggestions.

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 1 above)
Comment (drin) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:144]: I just broke this portion out into a separate function and tried to keep a consistent function name.

I can back out the re-organization though, I'll create a JIRA and try to move the changes to an appropriate PR

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 3 above)
Comment (drin) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:58]: I did this so that the structure would be consistent and thus easier to read.

think that at some point when we don't need it anymore we can just drop it. Since it is only internal to this source file, there shouldn't be any problem with the structure changing (I think).

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 4 above)
Comment (drin) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:87]: I made these changes to all instances of `ReadRel_LocalFiles_FileOrFiles`

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 3 above)
Comment (drin) [reply cpp/src/arrow/engine/substrait/relation_internal.cc]: I removed this as part of backing out the re-organization. I will include it in a subsequent PR.

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 1 above)
Comment (drin) [reply cpp/src/arrow/engine/substrait/relation_internal.cc]: I backed out the re-organization. I will move it to a new PR.

**Comment by drin** [comment]
rebased to accommodate ARROW-16855

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 7]
```diff
@@ -169,45 +157,90 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
...
+            ARROW_ASSIGN_OR_RAISE(auto file_info,
+                                  filesystem->GetFileInfo(item_uri.path()));
+
+            // push the FileInfo if it's a file; else, recurse into the directory
+            if (file_info.type() == fs::FileType::File) {
+              files.push_back(std::move(file_info));
+            } else if (file_info.type() == fs::FileType::Directory) {
+              RETURN_NOT_OK(DiscoverFilesFromDir(filesystem, item_uri.path(), files));
+            }
+
```
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:234]: nit: remove this empty line

**Review by vibhatha — COMMENTED** [review]
@drin Thank you for working on this. It looks good to me. 

@westonpace could you please take a look?

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 7 above)
Comment (drin) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:234]:
```suggestion
```

**Comment by vibhatha** [comment]
@drin One thing I forgot to mention, `test_substrait.py` also contains a few test cases skipping windows because of the same issue.

cc @westonpace should it be a follow up PR?

**Comment by drin** [comment]
I can enable those tests too... though, I guess I don't know if it's better to keep C++ updates and python updates separate? I can just add the [Python] tag to the name..

**Comment by vibhatha** [comment]
> I can enable those tests too... though, I guess I don't know if it's better to keep C++ updates and python updates separate? I can just add the [Python] tag to the name..

Some of the CIs are failing, apparently unrelated to this change AFAIU. Since the change we do is for Windows not sure if we have to wait till these are fixed.

**Comment by vibhatha** [comment]
One more thing, regarding https://github.com/apache/arrow/actions/runs/3055769050/jobs/4929199537#step:5:916

Here the linter fails because of an unused import. Let's remove that. Also it would be easier to avoid these by running, 
`archery lint --python --fix`

**Comment by drin** [comment]
lol, that's funny because I didn't add that import. I'll try to fix it tonight sometime

**Comment by drin** [comment]
oh I see, it was used for skipping, but once I removed the skips the package is no longer used. fair enough

**Comment by vibhatha** [comment]
> oh I see, it was used for skipping, but once I removed the skips the package is no longer used. fair enough

In case we miss, the linters can always help. 🙂

**Comment by pitrou** [comment]
@drin Can you rebase this?

**Comment by drin** [comment]
yep! I need to lint it too. I'll do that now before I forget

**Comment by drin** [comment]
rebased.

that was... not fun

**Comment by drin** [comment]
I was able to build and run substrait unit tests on this commit, so hopefully it is good now

**Comment by pitrou** [comment]
> rebased.
> 
> that was... not fun

Sorry :-(

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 8]
```diff
@@ -94,6 +94,20 @@ Status CheckRelCommon(const RelMessage& rel) {
   return Status::OK();
 }
 
+// Other helper functions
+Status DiscoverFilesFromDir(std::shared_ptr<fs::LocalFileSystem>& local_fs,
+                            std::string dirpath, std::vector<fs::FileInfo>& rel_fpaths) {
```
Comment (pitrou) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:99]: As per the coding conventions, we want to avoid non-const references.
Immutable args should be passed by value or const-ref.
Mutable args should be passed by pointer.
```suggestion
Status DiscoverFilesFromDir(const std::shared_ptr<fs::LocalFileSystem>& local_fs,
                            const std::string& dirpath, std::vector<fs::FileInfo> *rel_fpaths) {
```

File: cpp/src/arrow/util/uri_test.cc [hunk 9]
```diff
@@ -267,6 +268,7 @@ TEST(Uri, FileScheme) {
   auto check_with_host = [&](std::string uri_string, std::string host,
                              std::string path) -> void {
     ASSERT_OK(uri.Parse(uri_string));
+    ASSERT_TRUE(uri.is_file_scheme());
```
Comment (pitrou) [inline cpp/src/arrow/util/uri_test.cc:271]: For completeness, also add a test where this returns false?

File: python/pyarrow/tests/test_substrait.py [hunk 10]
```diff
@@ -40,9 +39,6 @@ def _write_dummy_data_to_disk(tmpdir, file_name, table):
     return path
 
 
-@pytest.mark.skipif(sys.platform == 'win32',
-                    reason="ARROW-16392: file based URI is" +
-                    " not fully supported for Windows")
```
Comment (pitrou) [inline python/pyarrow/tests/test_substrait.py:45]: Neat!

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 11]
```diff
@@ -210,43 +197,93 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
...
+            // Let the filesystem API decide for us if the URI is a file or a directory
+            ARROW_ASSIGN_OR_RAISE(auto file_info,
+                                  filesystem->GetFileInfo(item_uri.path()));
+
+            // push the FileInfo if it's a file; else, recurse into the directory
+            if (file_info.type() == fs::FileType::File) {
+              files.push_back(std::move(file_info));
+            } else if (file_info.type() == fs::FileType::Directory) {
+              RETURN_NOT_OK(DiscoverFilesFromDir(filesystem, item_uri.path(), files));
+            }
```
Comment (pitrou) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:275]: Should probably also handle the `FileInfo::Unknown` and `FileInfo::NotFound` cases (by erroring out perhaps?).

`FileInfo::Unknown` can be got for non-regular files such as pipes, sockets etc.
`FileInfo::NotFound` you probably understand :-)

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 12]
```diff
@@ -94,6 +94,20 @@ Status CheckRelCommon(const RelMessage& rel) {
...
+// Other helper functions
+Status DiscoverFilesFromDir(std::shared_ptr<fs::LocalFileSystem>& local_fs,
+                            std::string dirpath, std::vector<fs::FileInfo>& rel_fpaths) {
+  // Define a selector for a recursive descent
+  fs::FileSelector selector;
+  selector.base_dir = dirpath;
+  selector.recursive = true;
+
+  ARROW_ASSIGN_OR_RAISE(auto file_infos, local_fs->GetFileInfo(selector));
+  std::move(file_infos.begin(), file_infos.end(), std::back_inserter(rel_fpaths));
```
Comment (pitrou) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:106]: This will also give you subdirectories, is that what you want or should you filter them out?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks a lot for the update. Just some minor comments below.

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 12 above)
Comment (drin) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:106]: I think Weston must have misunderstood me then. I had code for omitting subdirectories, so I can just put that back in

Comment (pitrou) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:106]: Ha, sorry :-)

**Comment by drin** [comment]
rebased again.

Also addressed review feedback

**Comment by pitrou** [comment]
A test fails on Windows CI, could you take a look?

**Comment by drin** [comment]
tricked myself, which probably means the tests are pretty good. I tried doing copy paste and changing the scheme, but that doesn't work for windows because it drops the leading `/` if the URI is a file scheme and a drive letter. So, for a non file scheme I had to add a leading `/` to the expected value

**Comment by drin** [comment]
rebased to pick up style fixes

**Review by pitrou — APPROVED** [review]
This looks good to me from my limited knowledge of Substrait.

**Comment by pitrou** [comment]
@westonpace Would you like to give this a final review?

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 13]
```diff
@@ -96,6 +96,25 @@ Status CheckRelCommon(const RelMessage& rel) {
   return Status::OK();
 }
 
+// Other helper functions
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:99]:
```suggestion
```
Not sure how useful this comment is

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 14]
```diff
@@ -96,6 +96,25 @@ Status CheckRelCommon(const RelMessage& rel) {
...
+                            std::vector<fs::FileInfo>* rel_fpaths) {
+  // Define a selector for a recursive descent
+  fs::FileSelector selector;
+  selector.base_dir = dirpath;
+  selector.recursive = true;
+
+  ARROW_ASSIGN_OR_RAISE(auto file_infos, local_fs->GetFileInfo(selector));
+  for (auto& file_info : file_infos) {
+    if (file_info.IsFile()) {
+      rel_fpaths->push_back(file_info);
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:111]:
```suggestion
      rel_fpaths->push_back(std::move(file_info));
```

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 15]
```diff
@@ -120,9 +140,8 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
                               FromProto(read.filter(), ext_set, conversion_options));
       }
 
+      // NOTE: scan_options->projection is not used by the scanner and thus can't be used
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:143]:
```suggestion
```
This isn't really true

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 16]
```diff
@@ -212,43 +204,102 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
...
+          default:
+            RETURN_NOT_OK(item_uri.Parse(item.uri_path_glob()));
+            break;
+        }
+
+        // Validate the URI before processing
+        if (!item_uri.is_file_scheme()) {
+          return Status::NotImplemented("substrait::ReadRel::LocalFiles item (",
+                                        item_uri.ToString(),
+                                        ") with non-filesystem scheme (file:///)");
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:231]:
```suggestion
                                        ") with non-file scheme (file:///)");
```
Technically, `s3://` is an S3 "filesystem".

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 17]
```diff
@@ -212,43 +204,102 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
...
+
+        switch (item.file_format_case()) {
+          case substrait::ReadRel::LocalFiles::FileOrFiles::kParquet:
+            format = std::make_shared<dataset::ParquetFileFormat>();
+            break;
+          case substrait::ReadRel::LocalFiles::FileOrFiles::kArrow:
+            format = std::make_shared<dataset::IpcFileFormat>();
+            break;
+          default:
+            // TODO: maybe check for ".feather" or ".arrows"?
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:254]:
```suggestion
```
We aren't checking file extensions here so no need.

**Review by westonpace — APPROVED** [review]
A few minor commenting nits but looks good otherwise

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 16 above)
Comment (drin) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:230]: mmm.. I don't think S3 is a filesystem, but changing `filesystem` back to `file` here actually makes more sense

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 17 above)
Comment (drin) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:253]: sounds good

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 16 above)
Comment (drin) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:230]: I changed `with non-file scheme` to `does not have file scheme`  so that it's also clear that the `(file://)` portion part *is* a file scheme (just fixing grammar ambiguity)

File: cpp/src/arrow/util/uri.h [hunk 18]
```diff
@@ -45,6 +45,9 @@ class ARROW_EXPORT Uri {
   /// explicit scheme.
   std::string scheme() const;
 
+  /// Convenience function that returns true if the scheme() is "file"
+  bool is_file_scheme() const;
```
[the remark this answers is not in the document]
Comment (drin) [reply cpp/src/arrow/util/uri.h:49]: yeah, I thought it was interesting that there is an `is_file_uri_` property, but since there is a `port()` function that returns the `port_` property, I figured I would just keep consistent.

**Comment by pitrou** [comment]
CI passed on @drin 's fork.
